### PR TITLE
added --reserved-cpus kubelet command option

### PR DESCRIPTION
**Pull request by jianzzha**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
>
> /kind api-change
> /kind bug
> /kind cleanup
> /kind design
> /kind documentation
> /kind failing-test

/kind feature
> /kind flake

**What this PR does / why we need it**:

Currently when --system-reserved or --kube-reserved is used, kubelet will reserve system resources and keep them from being used by user containers. What CPUs are reserved via these options are not easy to determine by the end users, even though the CPU selection algorithm make the selection in deterministic way.

For telco NFV use cases, when using static cpu manager policy, knowing explicitly what CPUs are reserved for system and what CPUs are used for user containers is very important. For example, if isolcpu are going to be defined on the kernel cmd line and used for NFV containers, they really should be used only by the NFV containers and not by the system threads. Therefore, providing an option to explicitly define the system reserved CPU list will make the planning job much easier and straightforward for such telco use cases.

This new cmd line option, --reserved-cpus, is to explicitly define the the cpu list that will be reserved for system. For example, if --reserved-cpus=0,1,2,3 is specified, then cpu 0,1,2,3 will be reserved for the system. On a system with 24 CPUs, the telco user may specify isolcpus=4-23 for the kernel option and use CPU 4-23 for the NFV containers.

When --reserved-cpus is specified together --system-reserved or --kube-reserved, the explicit CPU list specified by --reserved-cpus will supercede the CPUs requested via --system-reserved or --kube-reserved.

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
A new kubelet command line option, --reserved-cpus, is introduced to explicitly define the the CPU list that will be reserved for system. For example, if --reserved-cpus=0,1,2,3 is specified, then cpu 0,1,2,3 will be reserved for the system.  On a system with 24 CPUs, the user may specify isolcpus=4-23 for the kernel option and use CPU 4-23 for the user containers.
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

**Comment by klueska** [comment]
I'm not opposed to this change in principle. However, I'd like to understand a bit better exactly how this will interact with `--system-reserved` and `--kube-reserved`.

You mention:
>> When --reserved-cpus is specified together --system-reserved or --kube-reserved, the explicit CPU list specified by --reserved-cpus will supercede the CPUs requested via --system-reserved or --kube-reserved.

Is it enough to just "supercede" these values? Can someone more knowledgable about `--system-reserved` and `--kube-reserved` chime in here?

/cc @ConnorDoyle @vishh

**Comment by mattjmcnaughton** [comment]
Hi @jianzzha !

Thanks for your thoughts here :) I defer to @derekwaynecarr and @dchen1107 here, but my gut says this is initiative would be best served by creating a [Kubernetes Enhancement Proposal](https://github.com/kubernetes/enhancements), especially given that it includes an api change.
The link I included should have tips for starting a KEP.

**Comment by jianzzha** [comment]
> I'm not opposed to this change in principle. However, I'd like to understand a bit better exactly how this will interact with `--system-reserved` and `--kube-reserved`.
> 
> You mention:
> 
> > When --reserved-cpus is specified together --system-reserved or --kube-reserved, the explicit CPU list specified by --reserved-cpus will supercede the CPUs requested via --system-reserved or --kube-reserved.
> 
> Is it enough to just "supercede" these values? Can someone more knowledgable about `--system-reserved` and `--kube-reserved` chime in here?
> 
> /cc @ConnorDoyle @vishh

The reserved CPU count overwrite happens at early phase, at the command line option processing time, from server.go:655. This way the behavior of how kubelet reports cpu resource number to the scheduler doesn't change (the additional code doesn't need to do anything on that regard). The only thing changed is how reserved cpu is assigned.

It works well in the test. The master sees the right cpu count deducted; the kubelet keep these cpus from assigned by the static policy.

File: cmd/kubelet/app/options/options.go
Comment (dustinblack) [inline cmd/kubelet/app/options/options.go]: Nit: "A comma-separated list of CPUs or CPU ranges in ASCII decimal..."

**Comment by derekwaynecarr** [comment]
@jianzzha - can you update the existing topology manager KEP here https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/0035-20190130-topology-manager.md with this update and use case?  Can we discuss in next week SIG node call as well?

File: staging/src/k8s.io/kubelet/config/v1beta1/types.go
Comment (derekwaynecarr) [inline staging/src/k8s.io/kubelet/config/v1beta1/types.go]: The field requires validation.

File: cmd/kubelet/app/server.go
Comment (derekwaynecarr) [reply cmd/kubelet/app/server.go]: So by overriding, this will ensure that you don’t reserve more than capacity with normal validation... the error message may be confusing as it will complain about system-reserved...

**Review by derekwaynecarr — CHANGES_REQUESTED** [review]
I am fine with this approach.  It enables users who desire explicit control to use it, and those that are comfortable with the default heuristic to work as normal.  I personally see little value in splitting kube-reserved from system-reserved for cfs sharing in this deployment as those users are probably running with real-time and will instead just want all things tied to specific cores.  If you add validation, I am fine with the change, and we can discuss in next week sig node to see if there is any disagreement.

File: cmd/kubelet/app/options/options.go
Comment (jianzzha) [reply cmd/kubelet/app/options/options.go]: > Nit: "A comma-separated list of CPUs or CPU ranges in ASCII decimal..."
"A linux CPU list formatted string comprised of comma-separated list of CPUs or CPU ranges" sounds ok?

Comment (jianzzha) [reply cmd/kubelet/app/options/options.go]: sounds good

File: cmd/kubelet/app/server.go
Comment (jianzzha) [reply cmd/kubelet/app/server.go]: Ack. Actual deployments seems just have the systemd unit via system.slice. I see your point, we need to have a further discuss in a design meeting

File: cmd/kubelet/app/server.go
Comment (jianzzha) [reply cmd/kubelet/app/server.go]: Ack, will fix the error message.

File: staging/src/k8s.io/kubelet/config/v1beta1/types.go
Comment (jianzzha) [reply staging/src/k8s.io/kubelet/config/v1beta1/types.go]: good reminder. will add it in two place: validation.go and server.go. In validation phase check it is a valid cpu list string, also make sure user does not intent to use  --kube-reserved-cgroup nor  --system-reserved-cgroup with the specific cpulist; in server.go make sure the cpulist is supported by the machine. If none of these condition meet, always take the old behavior

Comment (jianzzha) [reply staging/src/k8s.io/kubelet/config/v1beta1/types.go]: updated

**Comment by Levovar** [comment]
just to double check: what happens when I let's say configure --reserved-cpus=0,12; and CPU-Manager is either disabled, or it is enabled but policy is set to "none"?
how will the cpuset.cpus of my container look like on that 24 core system?

**Comment by jianzzha** [comment]
> just to double check: what happens when I let's say configure --reserved-cpus=0,12; and CPU-Manager is either disabled, or is enabled but policy is set to "none"?
> how will the cpuset.cpus of my "global shared" CPU pool using container look like on that 24 core system?

all the existing settings, --system-reserved, --kube-reserved, and this new setting only works when static cpu policy manager is used. If no policy specified (default to none), nothing will happen with any of these settings.

**Comment by Levovar** [comment]
> all the existing settings, --system-reserved, --kube-reserved, and this new setting only works when static cpu policy manager is used. If no policy specified (default to none), nothing will happen with any of these settings.

that was my understanding as well, just wanted to confirm :)
so why I'm asking this is because I recently wrote a KEP for this change. it is kind of obsolete now considering you already did the PR, but I did argue in my proposal that this new flag should be independent from all other node settings (unlike current way of defining reservations)

sure, current solution is okay for the use-case when the latency sensitive workload is on top of Kubernetes, but it does not address the cases when it is only **next** to K8s. A PTP systemd service, an OVS-DPDK switch, a VDU _virtual machine_.
You get my point. Not all TelCo use-cases -admittedly the use-cases behind your PR- use the static policy manager. In fact, I would even argue _most of them don't use it_.

So even considering this is not the current implementation, IMO the meaning of a newly introduced flag such as this really should be "you cannot touch these cores, dear Kubernetes".
Under any circumstances.

**Comment by jianzzha** [comment]
> So even considering this is not the current implementation, IMO the meaning of a newly introduced flag such as this really should be "you cannot touch these cores, dear Kubernetes".

This new flag in this PR is nothing more but to make the reserved cpuset explicit under static cpu policy. It's intended to be a increment change.

what you described is a different user story. It needs to be handled but probably with a larger context. Can you post that kep? @derekwaynecarr @ConnorDoyle might already have that big picture...

**Comment by Levovar** [comment]
> This new flag in this PR is nothing more but to make the reserved cpuset explicit under static cpu policy. It's intended to be a increment change.
> 
> what you described is a different user story. It needs to be handled but probably with a larger context. Can you post that kep? @derekwaynecarr @ConnorDoyle might already have that big picture...

before you misunderstand: I'm definitely not trying to push you to do anything outside your original scope, just wanted to raise my point related to it :)
incremental changes is always the way to go, I totally agree!

yep, I might still post the KEP but update it first to refer to your PR as the base, or first phase if you are okay with that

**Comment by derekwaynecarr** [comment]
This is largely localized to just the kubelet configuration and doesn't span components.  Its not strictly required that we need a kep.

/ok-to-test

**Comment by atheurer** [comment]
> > This new flag in this PR is nothing more but to make the reserved cpuset explicit under static cpu policy. It's intended to be a increment change.
> > what you described is a different user story. It needs to be handled but probably with a larger context. Can you post that kep? @derekwaynecarr @ConnorDoyle might already have that big picture...
> 
> before you misunderstand: I'm definitely not trying to push you to do anything outside your original scope, just wanted to raise my point related to it :)
> incremental changes is always the way to go, I totally agree!
> 
> yep, I might still post the KEP but update it first to refer to your PR as the base, or first phase if you are okay with that

Hello.  I work with @jianzzha on this stuff at Red Hat.  I wanted to give a little more context on how we wanted to use this feature, and why I think your point about the non-static cpu policy manager is just as important.

When we can "lock-down" what kublet does not use, we can much more easily tune the host to ensure other tasks do not interrupt any of the containers.  That type of tuning is usually done at boot-time, and so having a static list of CPUs we know kublet will never touch is important.  For example, we may choose to move all device interrupt processing to only the cpus in "--reserved-cpus", or we may update systemd's CPUaffinity to start with that same cpu list.  There are other things we can do like moving some RCU kernel threads to only run in that CPU list, reducing further the preemptions we could get on those CPUs.  For convenience, we usually do this with https://github.com/redhat-performance/tuned/tree/master/profiles/cpu-partitioning and other tuned profiles, but it does not require that.

When we do this, we size "--reserved-cpus" to be just enough CPUs to handle that work.  Those cpus could get quite busy, especially when all device interrupt processing moves to them.  For this reason, I do think it's a good idea to allow containers not using static cpu policy manager to [at least optionally] also not run on those cpus.  If we don't , it could be bad for those system services we moved there, as well as potentially bad for the containers. Granted, the host cpu scheduler should try to avoid unfair conditions like this, but I think it's better to have the control if you want it.

BTW, one of our tests is for 0-loss networking with DPDK.  When we have this type of isolation, plus the host tuning, we can achieve ~60 Mpps with 2 CPUs with no packet loss (we let this run for 2 hours with absolutely no packet loss to be fairly confident it is consistent).  Without this type of tuning, the interruptions from non-container work (system tasks, kernel, interrupts, etc), drops this very significantly.

**Comment by lmdaly** [comment]
I am in agreement that the reserved cpus as well as the kube/system reserved should be something that kubernetes doesn't touch and they should be entirely removed from the cpus_allowed list for all containers in pods managed by K8s.

However I do think that is a wider piece of work that this is a step toward, so in terms of this PR I think it lgtm as is (Once tests pass etc.). 

I believe @Levovar has a KEP in the works for this extra work that we could discuss how to do such a thing in.

**Comment by jianzzha** [comment]
> I am in agreement that the reserved cpus as well as the kube/system reserved should be something that kubernetes doesn't touch and they should be entirely removed from the cpus_allowed list for all containers in pods managed by K8s.
> 
> However I do think that is a wider piece of work that this is a step toward, so in terms of this PR I think it lgtm as is (Once tests pass etc.).
> 
> I believe @Levovar has a KEP in the works for this extra work that we could discuss how to do such a thing in.

Yep, I agree with the above assessment. I will fix the CI error and proceed.

**Comment by Levovar** [comment]
3 +1s are enough for me :)
let's continue the discussing the big picture in https://github.com/kubernetes/enhancements/pull/1319

**Comment by jianzzha** [comment]
@derekwaynecarr can you please take a look of the validation added per your request. thanks!

**Comment by derekwaynecarr** [comment]
thank you for adding the validation and tests.

Agree with @lmdaly that this is a good step forward, and makes what was a heuristic actually explicit.  This was important where a heuristic was insufficient.

/approve
/lgtm

File: pkg/kubelet/apis/config/validation/validation.go
Comment (liggitt) [inline pkg/kubelet/apis/config/validation/validation.go]: does this parse not catch the invalid characters checked for above?

**Comment by liggitt** [comment]
once the config reflects the list structure of the option, the config API type change lgtm

File: cmd/kubelet/app/server.go
Comment (jianzzha) [reply cmd/kubelet/app/server.go]: yes this is intentional, if it makes this far, I wanted it to fall back to the old behavior (the none-explicit way) as if this option was not defined. I debated about this myself and will change it.

**Comment by Levovar** [comment]
@derekwaynecarr absolutely!

I decided to still go ahead with the KEP to highlight what additional, related changes we might need to support all use-cases.
But I also referenced this PR at the same time as a fine "alpha" stage, or "step1".

We can discuss how -if at all- to proceed post-merge of this PR under the KEP, or an enhancement issue!

File: cmd/kubelet/app/options/options.go
Comment (jianzzha) [reply cmd/kubelet/app/options/options.go]: ack

File: pkg/kubelet/apis/config/validation/validation.go
Comment (jianzzha) [reply pkg/kubelet/apis/config/validation/validation.go]: consolidated the validation into one place. pls check again.

File: pkg/kubelet/apis/config/types.go
Comment (liggitt) [reply pkg/kubelet/apis/config/types.go]: this comment is still outstanding

File: pkg/kubelet/apis/config/validation/validation.go
Comment (liggitt) [reply pkg/kubelet/apis/config/validation/validation.go]: if you're only going to do the validation in one place, checking for well-formed options is better to do in validation.go, so that we can validate a config without actually starting a kubelet

File: pkg/kubelet/apis/config/types.go
Comment (jianzzha) [reply pkg/kubelet/apis/config/types.go]: the reason string is used rather than []string is because cpuset.Parse take string as input; also "-" is expected, in addition to just ","

Comment (liggitt) [reply pkg/kubelet/apis/config/types.go]: > also "-" is expected, in addition to just ","

iiuc, the - denotes a range, so `0,1,3-4` would parse to ["0","1","3-4"], right?

Comment (liggitt) [reply pkg/kubelet/apis/config/types.go]: looking at `Parse`, the string format is an [externally accepted format](http://man7.org/linux/man-pages/man7/cpuset.7.html#FORMATS). If that's the case, using that format in the config seems ok.

File: pkg/kubelet/apis/config/validation/validation.go
Comment (jianzzha) [reply pkg/kubelet/apis/config/validation/validation.go]: updated as we discussed.

File: pkg/kubelet/apis/config/validation/validation.go
Comment (liggitt) [inline pkg/kubelet/apis/config/validation/validation.go]: shouldn't we include the parse error in the validation error message?

Comment (jianzzha) [reply pkg/kubelet/apis/config/validation/validation.go]: updated

Comment (jianzzha) [reply pkg/kubelet/apis/config/validation/validation.go]: @liggitt ok to proceed?

File: pkg/kubelet/apis/podresources/v1alpha1/api.pb.go
Comment (liggitt) [inline pkg/kubelet/apis/podresources/v1alpha1/api.pb.go]: this change is unexpected... what regenerated this?

Comment (jianzzha) [reply pkg/kubelet/apis/podresources/v1alpha1/api.pb.go]: thanks for noticing it. must be one of the tools under hack that was used to correct the CI error. Let me revert this file and see CI error is that.

File: staging/src/k8s.io/code-generator/_examples/apiserver/openapi/zz_generated.openapi.go
Comment (liggitt) [inline staging/src/k8s.io/code-generator/_examples/apiserver/openapi/zz_generated.openapi.go]: these generated changes are also unexpected

File: staging/src/k8s.io/sample-apiserver/pkg/generated/openapi/zz_generated.openapi.go
Comment (liggitt) [inline staging/src/k8s.io/sample-apiserver/pkg/generated/openapi/zz_generated.openapi.go]: same here

File: staging/src/k8s.io/code-generator/_examples/apiserver/openapi/zz_generated.openapi.go
Comment (jianzzha) [reply staging/src/k8s.io/code-generator/_examples/apiserver/openapi/zz_generated.openapi.go]: these are changed by hack/update-bazel.sh and hack/update-codegen.sh, or the CI will fail by hack/verify-bazel.sh and hack/verify-codegen.sh

Comment (jianzzha) [reply staging/src/k8s.io/code-generator/_examples/apiserver/openapi/zz_generated.openapi.go]: so I reverted these auto generated files, local check using hack/update-bazel.sh and hack/update-codegen.sh fails, but upstream looks fine. So yeah they are not required to pass the upstream CI. Thanks for pointing this out.

**Comment by liggitt** [comment]
/test pull-kubernetes-dependencies

managing the queue around the go1.13 bump, see https://groups.google.com/forum/#!topic/kubernetes-dev/Yyka3G2ebXE

**Pull request merged**
